### ISSUE TITLE
BUber: Proposal for instant permissionless Taxi service on Nostr

### ISSUE THREAD

**Issue by arcbtc** [fiatjaf/nostr#21]
Uber works well, apart from the Uber company.

BUber uses Nostr https://github.com/fiatjaf/nostr network to connect taxis and clients, in a permissionless way.

### Setup:

User opens software and selects whether they are a *taxi* or *customer* (either choice issues Schnoor keypair, or they can add their own pre-existing keypair)

### Workflow:

**Taxi:** Using their private key, the *taxi* publishes their `geolocation`, `rate`, and `availability` to relays.

**Customer:** Software recieves all *taxis* in the location of the *customer* and lists their `rate`, `availability` and `reputation` (if the taxi has `reputation` https://github.com/fiatjaf/nostr/issues/20).

**Customer** Using their private key, the *customer* publishes a request for a particular *taxi* and where they want to go.

**Taxi:** The *taxi* receives the *customers* request, and `reputation`, if the *customer* has `reputation` https://github.com/fiatjaf/nostr/issues/20

**Taxi:** Using their private key, the *taxi* accepts the job. `Geolocation` data is sent every few seconds to the *customer*, so they can track.

**Customer/Taxi:** Service happens, `reputation` is given at the end of the transaction.

> Note: I will build an https://github.com/lnbits/lnbits extension that acts as a Nostr client software for *taxi* and *customer*, but any client software for BUber will perform the same functions

**Comment by Entryist** [comment]
ACK - Looks good.

Minor point, perhaps consider swapping out "Taxi" terminology for "Driver" as this setup can be used for a variety of services requiring service providers that are mobile.

For extensibility I think it's important for client software to be able to attach arbitrary `metadata` such as service type (ride, delivery), vehicle type (car, bike, truck, bus), vehicle capacity, handicapped support, and anything else maybe not relevant in all use cases but extremely important for some customers. Important also for client-specific reputation layers that can't be expressed as an integer like Arcade City guild membership. Provides an incentive for client developers to work together to recognize each others' metadata, form alliances or common standards, and other uses we can't currently envision.

**Comment by arcbtc** [comment]
Aye, very good points!

**Comment by fiatjaf** [comment]
I think we should restrict this proposal to taxi service. Other services will need other kinds of metadata, other clients, and thus they need other NIPs with their own specs. Better not try to mangle everything in a single place.

Also if we need extra metadata here, like the type of vehicle, it's better to specify it in the NIP -- even if marked as optional.

**Comment by fiatjaf** [comment]
From @Entryist on https://t.me/nostr_protocol/6015:

> Christopher David, [04.09.21 13:47]
> A few things we've communicated to drivers:
> - Drivers can set their own rates
> - Drivers can build their own recurring customer base
> - Drivers can form guilds - have explored using Sphinx's "tribes" for this, or will reimplement our own version if needed
> - Riders can throw a request to 1) a specific/favorite driver, 2) all drivers in a guild (bc they like their policies), 3) or all drivers in an area who meet specific criteria (filter by time on platform, # of reviews, dog-friendly, whatever)
> 
> Not everything needs to be implemented all at once, and happy to get something basic in place sooner than later

I like this idea, although it's not very different from @arcbtc's proposal. The guilds thing is very interesting, I can think of some ways of implementing it, but I think v1 of this protocol should probably not bother with them.

**Comment by arcbtc** [comment]
> I think we should restrict this proposal to taxi service. Other services will need other kinds of metadata, other clients, and thus they need other NIPs with their own specs. Better not try to mangle everything in a single place.
> 
> Also if we need extra metadata here, like the type of vehicle, it's better to specify it in the NIP -- even if marked as optional.

I think with a well thought out nip, tags can serve a vast spectrum of `geolocated services`

**Comment by Entryist** [comment]
I like "Drivers" as the focus for the first implementation. Our client app supports services categorized as "ride", "delivery", "other" ([example screenshot](https://twitter.com/ArcadeCityHall/status/1402647814300110850/photo/1)) with all drivers treated similarly regardless of what location-based service they provide, with details of service worked out in chat between service provider and customer.

The terminology may not matter, as clients can display what they want, and any needed metadata can probably be put in the BIP-01 tags. 

But I would recommend avoiding the word "taxi" unless you don't mind a higher likelihood for scrutiny from transportation regulators that regulate taxis. I think "Driver" is the right amount of generic.

**Comment by arcbtc** [comment]
> service details worked out in chat between service provider and customer
So we could include DM notes between seller and customer https://github.com/fiatjaf/nostr/blob/master/nips/04.md

**Comment by fiatjaf** [comment]
I think DM notes would end up playing a big role on everything.

**Comment by misovan** [comment]
I stumbled around this idea only now when playing with branle. 1.5 years ago I did (as a kind of personal code challenge) similar service focused on park space sharing - [.](https://surfcity.app). Apart from card / sepa payments and payouts it did include full support of lightning payments and payouts - using lnurl-w. After few months of usage by some enthusiasts in my city it did not really take off. Nevertheless, even centralized by nature some ideas might work for buber-like services as well:

- Same as buber identities are in essence key pairs, pk is kept on device only
- Any space buyers personal info is kept on device and encrypted after order for parking space is paid, using buyers privkey and owners pubkey, so not readable on server (car type and plate, phone number to resolve exceptions)
- Only space owner can decrypt this data on his device - as it was encrypted using shared DH key
- Attestations of space ownership are done via digital signatures by other space owners with attested ownership (in person, via in app QR code scan - not fully bullet proof but worked well enough in residential communities)

As I've understood, relays currently process signed but not encrypted information and this could be the way not to transmit private info through relays and to introduce one possible concept for communities based attestations.

**Comment by fiatjaf** [comment]
I'm somewhat afraid of launching something like BUber because I think the reputation scheme must be very good to prevent horrible abuses. It's also tricker to do because of location etc.

Something like surfcity seems less problematic, am I right? I think we should do it and copy your code and your protocols. What do you think?

**Comment by misovan** [comment]
I understand the concerns. Regarding the risks involved, sharing parking space is nowhere near to uber-like service.
However it seems that low risk means as well low potential income. In this case it was likely not motivating enough to enroll  critical mass even in suitable areas with mostly private parking spaces. In some areas it simply morphed into fb groups of neighbors asking and providing their spaces for free.

Centralized service means most likely incoporation is needed - e.g. to release on iOS, to accept fiat payments or gain basic trust of users - which is of course not an issue as such in normal scenarios. Lightning does not bring real p2p across ordinary users, because they have no clue. So I used lnbits server to manage their accounts and withdrawals on top of dedicated lnd. Neat function is that they could withdraw BTC anytime - whereas if they chose EUR, they had to wait for the end of month to receive. But once again - for small amounts people do not care.

From above experience I am a little bit reserved for now to decentralization / privacy / BTC payments as motivational enough for mass and already working use case. 

To conclude I like the idea of nostr-like open messaging protocol with added specific business logic and payments layers mostly for cases where motivation is high alongside with high barriers to enroll for traditional service for whatever reason. 
Perhaps for now good next step might be to solve how to design and plug in those additional layers in some generic way, not stepping into exact use cases. Times might come, that it will be needed in some countries to live normal life :( Of  course I might be plain wrong.

I can of course share the code (react native app, lnbits server proxy and withdrawal logic, bunch of helper functions running on firebase to make notifications) if it would be of any use for you - I was more then delighted to play with a lot of your projects 👍

**Comment by trosel** [comment]
> because I think the reputation scheme must be very good to prevent horrible abuses

I agree. In fact, I don’t think a reputation scheme is enough for an uber type app. Ideally there would be a background check service before allowing people to sign up as drivers in the first place. Not full-proof, but competing background check services would be best.

**Comment by gourcetools** [comment]
This is cool idea!

**Comment by PatrickGeyer** [comment]
Couldn't a deposit/escrow service be required for this type of app as well as for marketplace/airbnb type apps?
Would be great if there was a spec/NIP for announcing escrow services to Nostr. I.e. announcing fee of escrow, types of payments that you escrow, and how you settle (hodl invoice for e.g.)

That would mean all these nostr apps could just scan the network for a suitable escrow when it's required. Greatly reduces the need for flawless reputation if trustworthy (yet non-custodial) escrow can be used.

Anyone working on this?
